huggingface_pipelines/speech.py

```diff
@@ -0,0 +1,150 @@
+            speech_ctx = SpeechInferenceParams(
+                data_file=self.config.data_file,
+                audio_root_dir=self.config.audio_root_dir,
+                audio_path_index=self.config.audio_path_index,
+                target_lang=self.config.target_lang,
+                batch_size=self.config.batch_size,
+                pad_idx=self.config.pad_idx,
+                device=self.config.device,
+                fbank_dtype=self.config.fbank_dtype,
+                n_parallel=self.config.n_parallel
+            )
+            speech_to_text_dp = self.speech_to_text_pipeline.build_pipeline(
+                speech_ctx)
+            transcriptions = []
+            with torch.inference_mode():
+                for batch in speech_to_text_dp:
+                    transcriptions.extend(batch)
+            logger.info("Audio transcribed successfully.")
+            return transcriptions
+        except Exception as e:
+            logger.error(f"Error transcribing audio: {e}")
+            raise
+
+    def process_batch(self, batch: Dict[str, Any], dataset: Dataset) -> Dataset:
+        """
+        Processes a single batch of data by transcribing audio and updating the dataset.
+
+        Args:
+            batch (Dict[str, Any]): A batch of data containing audio.
+            dataset (Dataset): The dataset to update.
+
+        Returns:
+            Dataset: The updated dataset.
+        """
+        for column in self.config.columns:
+            audio_data = batch[column]
+            transcribed_texts = self.transcribe_audio(audio_data)
+            dataset = dataset.add_column(
+                column + '_transcribed', transcribed_texts)
+        return dataset
+
+    def cache_results(self):
+        """
+        Caches the results to a JSON file.
+
+        The results are saved in a file named 'output_file_name_shard_{shard_id}.json'.
+        """
+        try:
+            file_name = f'{self.config.output_file_name}_shard_{self.config.shard_id}.json'
+            logger.info(f"Caching results to {file_name}...")
+            with open(file_name, 'w') as f:
+                json.dump(self.results, f)
+            logger.info("Results cached successfully.")
+        except Exception as e:
+            logger.error(f"Error caching results: {e}")
+
+    def cache_results_arrow(self):
+        """
+        Caches the results to an Arrow file.
+
+        The results are saved in a file named 'output_file_name_shard_{shard_id}.arrow'.
+        """
+        try:
+            file_name = f'{self.config.output_file_name}_shard_{self.config.shard_id}.arrow'
+            logger.info(f"Caching results to {file_name}...")
+            dataset = Dataset.from_dict({
+                "original": [result['original'] for result in self.results],
+                "transcribed": [result['transcribed'] for result in self.results],
+                "reference": [result['reference'] for result in self.results]
```

Comment:
we should not use self.results at all. 
My impression is that the caching on disk or in memory should not be the pipeline responsibility, i could be done outside on the dataset object directly.

Reply:
not that what you want here is probably implement at `dataset.map` level as well:
<img width="1045" alt="image" src="https://github.com/user-attachments/assets/6d7d58a8-253a-4df7-b884-646dfc5f7ea8">

Reply:
I removed this to adopt the more functional approach of caching results, these changes were made prior to our meeting.